### PR TITLE
riscv64: Consolidate conditional moves into one instruction 

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit removes the `IntSelect` and `SelectReg` pseudo-instructions
from the riscv64 backend and consolidates them into the `Select`
instruction. Additionally the `Select` instruction is updated to subsume
the functionality of these two previous instructions. Namely `Select`
now operates with `ValueRegs` to handle i128 and additionally takes an
`IntegerCompare` as the condition for the conditional branch to use.

This commit touches a fair bit of the backend since conditional
selection of registers was used in quite a few places. The previous
`gen_select_*` functions are replaced with new typed equivalents of
`gen_select_{xreg,vreg,freg,regs}`. Furthermore new `cmp_*` helpers were
added to create `IntegerCompare` instructions which sort-of match
conditional branch instructions, or at least the pnemonics they use.

Finally since this affected the `select` CLIF instruction itself I went
ahead and did some refactoring there too. The `select` instruction
creates an `IntegerCompare` from its argument to use to generate the
appropriate register selection instruction. This is basically the same
thing that `brif` does and now both go through a new helper,
`lower_int_compare`, which takes a `Value` and produces an
`IntegerCompare` representing if that value is either true or false.
This enables folding an `icmp` or an `fcmp`, for example, directly into
a branching instruction.

### CODE REVIEW

File: cranelift/codegen/src/isa/riscv64/inst.isle [hunk 1]
```diff
@@ -2684,25 +2639,92 @@
...
+(rule 1 (lower_int_compare val @ (value_type $I128))
+  (cmp_nez (rv_or (value_regs_get val 0) (value_regs_get val 1))))
+
+;; If the input value is itself an `icmp` we can avoid generating the result of
+;; the `icmp` and instead move the comparison directly into the `IntegerCompare`
+;; that's returned. Note that comparisons compare full registers so
+;; sign-extension according to the integer comparison performed here is
+;; required.
+(rule 2 (lower_int_compare (maybe_uextend (icmp cc a b @ (value_type (fits_in_64 in_ty)))))
+  (int_compare cc (zext a) (zext b)))
```
Comment (afonso360) [inline cranelift/codegen/src/isa/riscv64/inst.isle:2711]: Another small optimization here is that we could sign extend `IntCC::Equal` and `IntCC::NotEqual` for 32bit values. To take advantage of the same sign extend instruction as above.

**Review by afonso360 — APPROVED** [review]
This looks great to me! Thanks! A small suggestion below.

I also noticed that we are missing support for the `{u,s}{min,max}` instructions that I thought were implemented, but apparently not. I'll submit a follow up PR to this one adding them.

File: cranelift/codegen/src/isa/riscv64/inst.isle (hunk 1 above)
Comment (alexcrichton) [reply cranelift/codegen/src/isa/riscv64/inst.isle:2717]: Excellent point!

I swear I'm eventually getting back towards icmp (slowly!), and I thought this was a neat optimization in your original PR for icmp. I've added it here as well

Comment (afonso360) [reply cranelift/codegen/src/isa/riscv64/inst.isle:2717]: Even without `icmp` these refactors are very much appreciated!

**Pull request merged**